### DISCUSSION TITLE
SubTypes (not Alias)

### DISCUSSION THREAD

**Discussion by AnthonyVO** [dotnet/csharplang#6406] [Language Ideas]
I am sure that others have thought of this LONG before me, but I still don't see this feature implemented, and I believe it would be extremely valuable with the potential to save millions in losses due to failed calculations and debugging.  (e.g. $125 million lost Mars orbiter due to a missing metric conversion, or worse, patient overdose due to conversion between grams and grains)

Most developers will default to using the base types whenever they can because they are simple and they don't require the creation of heavier structures or classes with all of the scaffolding.  The BIG problem this creates is that the base types such as Int, Double and String carry no meaning and no real type information other than the base.  

This means that we have no way to tell the compiler that some types can't be converted to others without some explicit conversion.

The result is that it is extremely easy to accidentally pass a variable of one type (miles, height, pounds, or EMs) into a function expecting (meters, width, kilograms, or pixels), and there is no way that the compiler would be able to block that mistake and it is extremely easy to miss that as a developer.  It would only be caught by careful testing or expensive debugging.

This is also true of String values.  Many string values represent intermediate or processed values that presume special handling.  A password, sql where clause, GMT date string, or unsafe user input, are all different types of string values that should never be confused with each other but which are all valid string values.

In physics or mathematics, any calculation that resulted in a value that had no units would automatically be rejected as wrong.   

In c#, we have no simple way to specify a value's Unit or SubType.  The only option is to create a struct, class or enum.

I understand that we can't subclass the base types, so I propose that c# be extended to allow the creation of SubTypes that compile down to their base types for use at runtime and serialization.

The SubTypes would only be used during development and at compile time to block invalid conversions.  

For this reason, the concept of a SubType is stronger than that of an Alias.  The whole purpose is to get type checking while still being able to use the base types efficiently.

I imagine an implemention similar to:

```
subtype Kilometers is Double;
subtype Miles is Double;

public Kilometers ToKilometers(this Miles miles) 
{
   return (Kilometers)(miles * 1.609344);
}

Miles distanceToWork = 10;
var distanceToWorkInKM = distanceToWork.ToKilometers();

subtype EncodedHtml is string;

public EncodedHtml Encode(string html)
{  
    return (EncodedHtml)HttpUtility.HtmlEncode(html);
} 
```

I am not exagerating when I say that there is millions to be saved in errors and debugging.  This would help make development more  accurate and safe.

**Comment by HaloFour** [comment]
See: #5498

You could also do something like this with single-value struct records:

```cs
public record struct Miles(double value) {
    public static Kilometers ToKilometers() => new Kilometers(value * 1.609344);
}

public record struct Kilometers(double value) {
    public static Miles ToMiles() => new Miles(value / 1.609344);
}
```

The JIT is pretty good about eliding structs that wrap a single value like this.

**Reply by AnthonyVO** [reply]
Thanks @HaloFour, that is a good point, but it suddenly forces me to write a whole pile of extra code to access all of the normal functionality that comes for free with the base type.

Any work I do with the Miles value (now structure) must be specifically implemented within that structure.  I can't treat it like a Double type.  There is no ability to multiply, add or subtract unless I specifically implement it, or export the value and stuff it back in.

**Reply by CyrusNajmabadi** [reply]
> I can't treat it like a Double type. There is no ability to multiply, add or subtract unless I specifically implement it, or export the value and stuff it back in.

That's a virtue to me.  Not all functionality should necessarily be there.  And, if you do want all the operators of the underlying types, those coudl be generated by a source-generator.

**Reply by AnthonyVO** [reply]
But all of that extra code, generated or otherwise, is exactly why developers will revert to using simple base types.

It is just an integer value that holds a Kilometer type, so they just use an Int and lose the ability to have the compiler stop them when they try to do something invalid.

I have had that discussion with myself many, many times.  At what point do I recreate a base type in a struct or class simply to get that extra type checking while knowing I will need to build all of the support code?

I suggest `subtype Kilometers is Double;`

**Reply by CyrusNajmabadi** [reply]
> But all of that extra code, generated or otherwise, is exactly why developers will revert to using simple base types.

Why?  If it's generated, why would anyone care?

> I suggest subtype Kilometers is Double;

We're going to have to encode this as something anyways, and we'll have to do codegen anyways.  So why is using an SG to do just that (which can be done today) less palatable then us doing all that anyways, but having to wait 3-5 years until you can actually use it?

**Reply by AnthonyVO** [reply]
As indicated in the proposal.  At compile time the subtypes simply revert to their base types.  There is no extra source code that needs to be generated.  There is no extra runtime overhead.  The compiler would need to ensure it selects the correct extension methods, but otherwise, the subtypes are simply compiled out.  The base types remain the base types.  A string is serialized as a normal string.

The type-checking is done at development and compile time.

I guess I should have suggested it 3-5 years ago when I first thought of it.  My loss.

Failing to improve the language because there is a workaround (a source code generator), is still failing to improve the language.  The workaround creates an extra dependency that won't be available to everyone.

**Reply by CyrusNajmabadi** [reply]
>As indicated in the proposal. At compile time the subtypes simply revert to their base types. 

Then there can be no enforcement of whatever restrictions you want here.  

> The type-checking is done at development and compile time.

To be done at compile time, the information needs to be in the metadata as to what this is so it can have these extra rules.  Since we have to generate that anyways, it's unclear why we would not jsut generate a working system that strongly enforces safety here instead of weakly.

>  At compile time the subtypes simply revert to their base types. 

This would mean it would be trivial to subvert as anyone from any other language could pass you anything here.

> Failing to improve the language because there is a workaround (a source code generator), is still failing to improve the language. 

It's solving the problem in reasonable time frames.  That's why tehse tools exist.  So that not everything is bottlenecked by having to go through teh language on the timelines that language development happens at.

> The workaround creates an extra dependency that won't be available to everyone.

Same issue with a language change.  Not everyone is on the latest language.  SGs have the benefit of existing now and suitably being able to generate code that any consumer in the .net ecosystem can benefit from.

**Reply by AnthonyVO** [reply]
> Then there can be no enforcement of whatever restrictions you want here.

I may be treading on thin ice since I don't know the exact implementation, but I am thinking of something similar to Enums.  As far as I am aware, they are simple integers, and there is no runtime validation.  Yet Enum types are usable by extension methods and enable type checking.  The constant values and the assigned names are there simply to aid in development.

A subtype would operate in a similar manner.  It is only a development aid, the underlying types remain the underlying types.  

> This would mean it would be trivial to subvert as anyone from any other language could pass you anything here.

Understood.  Any public interface would need to validate its input.  Any method reading in data would need to validate its input.

**Reply by CyrusNajmabadi** [reply]
> I may be treading on thin ice

Not at all :)  All ideas are welcome here :)

**Reply by HaloFour** [reply]
@AnthonyVO 

> but I am thinking of something similar to Enums. As far as I am aware, they are simple integers, and there is no runtime validation. Yet Enum types are usable by extension methods and enable type checking. The constant values and the assigned names are there simply to aid in development.

There is no runtime validation, but they are still very special types as far as the runtime is concerned.  But even enums don't "inherit" all of the functionality of the underlying type.  For example, you can't multiply them together.

**Reply by Richiban** [reply]
> There is no ability to multiply, add or subtract unless I specifically implement it, or export the value and stuff it back in.

To be honest, I think this is something that you don't want to add in automatically. There's bound to be a delegation source generator that will add these if you want them.

As a counter example, think of a newtype called `Money` that wraps a decimal. You want to be able to add and subtract money values from one another, but what does it mean to _multiply_ or _divide_ money values? It doesn't make sense; a lot (most?) of the time you only want a subset of the base types members to be implemented, in which case you _should_ write them yourself.

**Reply by TahirAhmadov** [reply]
@Richiban `Money` is not a physical measurement, so it's probably not in the scope here. When it comes to physical measurements, most of them can be multiplied/divided by a _number_, and added/subtracted from each other. All that boilerplate is best left to a base type of some sort. Which is why I would like to see some kind of "struct inheritance", which basically becomes something like "mixins".

I tried looking into the SG approach, but it has downsides:
https://github.com/dotnet/csharplang/discussions/5897#discussioncomment-3142362

**Reply by TahirAhmadov** [reply]
More importantly, in addition to the common operations of mult/divide, "child structs" would then be able to define their own specific code.

**Reply by TahirAhmadov** [reply]
Also, with "mixins", we get interface implementations getting promoted, which also gets us static interface members, making this feature even more powerful. Basically, imagine you have all the `INumber` stuff, but you can use it in a type-safe manner with `Miles` and `Kilometers`, etc.

**Reply by HaloFour** [reply]
Delegating interface implementation is something also easily achievable via source generators.

**Comment by TahirAhmadov** [comment]
I really think we should seriously look into some kind of struct "inheritance"/"mixin" idea. It would solve these types of challenges perfectly.

**Reply by HaloFour** [reply]
I doubt they could ever really be more than a struct wrapper with delegating methods, something you could accomplish with a record struct and source generators already.  Given the amount of policy one would likely want to apply, such as filtering which members to be "promoted", it's likely the kind of thing that would involve so much special syntax to never be worth the ROI.

I'd rather see how the conversation around "roles" takes shape (pun intended), given the intent is to narrow the usage of an existing type.  At least that conversation has started and has interest from the team.

**Reply by TahirAhmadov** [reply]
As discussed elsewhere, the SG approach doesn't really work that well. It could achieve some of the benefits, but not all.

To me, the chief advantage of "mixins"/"struct inheritance" is that they are more in line with OOP, whereas the roles are more of a step in a different direction.

**Reply by CyrusNajmabadi** [reply]
> As discussed elsewhere, the SG approach doesn't really work that well. It could achieve some of the benefits, but not all.

What doesn't work well with the SG approach?

> To me, the chief advantage of "mixins"/"struct inheritance" is that they are more in line with OOP, whereas the roles are more of a step in a different direction.

I'm not sure what that means.  Can you elaborate?

**Reply by TahirAhmadov** [reply]
> What doesn't work well with the SG approach?

https://github.com/dotnet/csharplang/discussions/5897#discussioncomment-3142362

> I'm not sure what that means. Can you elaborate?

1, I'm not even sure how roles can solve the problem raised in the OP - units of measurements. 2, roles erode the unique identity of a type, we take X and pretend it's Y; to me, it sounds like the opposite of OOP; has a smell to it.

**Reply by HaloFour** [reply]
@TahirAhmadov 

That's increasing the complexity orders of magnitude to achieve what I consider to be pretty esoteric goals.

> roles erode the unique identity of a type, we take X and pretend it's Y; to me, it sounds like the opposite of OOP; has a smell to it.

Which is exactly what this proposal is suggesting as well, that the "subtype" is elided by the compiler.

**Reply by TahirAhmadov** [reply]
To be clear, I'm not advocating for the proposal in the OP; but I do think it's an indication there is a demand for a feature which solves it, and a whole bunch of other scenarios as well.

**Reply by HaloFour** [reply]
I agree, and the vast majority of those cases are solvable _today_ via a source generator.  I don't think the few that aren't, or the cases where the syntax isn't as "pretty" as it could be, warrants an entire language feature, especially one that wouldn't be likely to land for years even if it were to be championed today.

**Reply by CyrusNajmabadi** [reply]
>  discussed elsewhere, the SG approach doesn't really work that well

I took a look at your examples and I'm really not getting your issues. They seem solvable or extremely minor.

**Reply by AnthonyVO** [reply]
> especially one that wouldn't be likely to land for years even if it were to be championed today.

In answer I will quote:

> “The best time to plant a tree was 20 years ago. The second best time is now.” – Chinese Proverb.

If it is worth doing, then let's do it.  The fact that the feature won't land for a while is irrelevant.  Even if I have to wait till my retirement before I can enjoy it.  :)

**Reply by CyrusNajmabadi** [reply]
> The fact that the feature won't land for a while is irrelevant. Even if I have to wait till my retirement before I can enjoy it. :)

I strongly disagree with this.  Language design (and compiler implementation) is a zero sum game.  Everything we choose to do means a set of things we are not going to be able to do.  And so it's vital we focus on the things that are important and valuable for the ecosystem.

Conversely, 'Analyzers' and 'Source Generators' change the equation entirely.  They allow for many sorts of requests that would previously have only been possible in the language to now be done by the community.  They fundamentally changes things from zero-sum to a win-win since the c# team doesn't have to bear these costs, but hte community gets the end value.  It's why we built these two systems and why they're arguably the most important pieces of roslyn.  Any time something isn't that important, and can be done with tehse technologies instead, they should be used.

> If it is worth doing, then let's do it. 

Sure.  If you're interested, then def go ahead and do it with things like Source Generators.  You def have the passion here, so my recommendation would be to become the change you want to see in the world here :)

**Comment by huoyaoyuan** [comment]
This looks like the *unit of measure* feature of F#.
I did find this useful. If separated types are generated, there can be quite a lot to do for fitting, for example converters in serialization.

**Reply by ufcpp** [reply]
Yes. It is weird that Kilometers * Kilometers = Kilometers. The unit of measure is needed.

```cs
Kilometers x = 2;
Kilometers y = 5;
Kilometers z = x * y; // This should be km^2. The unit is mismatched.
```

**Reply by AnthonyVO** [reply]
> _unit of measure_ feature of F#.

https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/units-of-measure

You are correct, this is functionally similar to the _unit of measure feature in F#_.   To quote from the MS docs...

> "Floating point and signed integer values in F# can have associated units of measure, which are typically used to indicate length, volume, mass, and so on. **By using quantities with units, you enable the compiler to verify that arithmetic relationships have the correct units, which helps prevent programming errors.**"

The requested feature has two significant differences:
- It is not limited to numeric types.  Supporting the string type is crucial since there are many different types of strings that should never be confused.  e.g. Passwords vs User messages, Encoded Html vs unsafe user input, prepared sql expressions, specifically formatted data keys, etc...
- It has no extra functionality except to **hint** to the compiler that there are two (or more) distinct types that are not interchangeable except by explicit type casting.

Basically telling the compiler that two different types are **different** and should not be mixed unless the developer explicitly permits it.

**Reply by AnthonyVO** [reply]
As mentioned above.  This is also very similar to Enums.  Basically, the only difference between two different Enum **types** is the values that the compiler will permit you to assign, retrieve or compare to.   Note: for the sake of this argument I am ignoring the [Flags] variant.

Yet, as _simple_ as the Enum Type functionality is, I am sure most developers are more than happy to take advantage of that functionality to make their code more readable, maintainable and reliable.

**Reply by HaloFour** [reply]
As mentioned, Enums are a very special type to the runtime, which is treated differently by the runtime.  It has all of the additional metadata associated with it that allows it to have any kind of impact on the type system and to be recognized by any of the other languages that can target .NET.  Enums aren't exactly a simple solution here.

To achieve the type safety you want across assembly boundaries requires more than compiler changes.  It requires at least metadata to support the notion of the type.  The safest route would be to wrap the type in another type, as that's something that the runtime understands today.  Aside that you have attributes or modopts, which other languages have to be updated to understand and respect.  Even downlevel C# compilers would only recognize the type as the underlying type thus defeating the purported type safety.  Note that F# does this and it's trivial to break F#'s type safety in a lot of places when you call it from C#.

**Comment by dmchurch** [comment]
Agreed that this would be an extremely nice-to-have - I've been thinking very similar things to the OP, and I came here to make a very similar post (but now I don't have to).

I'll add is that as a language _user_, something like this is the syntax I'd expect to see, based on my experience of C#:

```csharp
public static class SafeMath
{
    public static double Sin<T>(double<T> angle) where T : INumericUnit, IConvertibleTo<Radian>
    {
        // return Math.Sin(angle); // <-- compile error, may not implicit-cast a typed unit to a base unit
        // return Math.Sin(angle.Value); // <-- compile error, may not take the value of a generic-typed unit
        return Math.Sin(angle.Value<Radian>()); // <-- specified type allows conversion to the base
        
    }
    public static double<Radian> Asin(double d)
    {
        // return Math.Asin(d); // <-- compile error, may not implicit-cast a base unit to a typed unit
        return (double<Radian>)Math.Asin(d); // Works, explicit-casting is one way to type a base unit
        return new double<Radian>(Math.Asin(d)); // Also works
    }
}
```

By seeing the `double<Radian>` radian syntax, I instantly know a number of things:

1. This is a type that is a double-precision floating point number, in the same way that a `List<int>` is, in fact, a list
2. There are probably some restrictions on converting between base `double` and `double<Radian>`, but I can probably insist on it by using a C-style typecast, just like with `IReadOnlyList<>` and its non-generic counterpart
3. I can probably also use `double<Degree>` and it will understand what I'm saying
4. I can add two `double<Radian>`s together and get a `double<Radian>` as a result, and it will just compile down to a single floating-point addition op

It also offers a fairly simple implementation direction: just flip the generic around. Under the hood, a `double<Radian>` is just a `Radian<double>` and can be used as such in other CLR languages or previous versions of C#, so backwards compatibility isn't lost. `Radian<T>`, then, is just a generic type with some explicit requirements that allow it to be used as a "numeric unit type":

1. It must be a readonly struct with a single generic argument (T) constrained to being of "numeric type"
2. It must have exactly one field member, which must be public, readonly, and of the type T
3. It must have a single-parameter constructor that takes a type T
4. It probably has a type attribute that says it's a unit type, a la Enum's `[Flags]`

So that's no changes to the JIT (by the time it gets to IL it's already in flipped form, much like extension methods), no lexical changes to the C# language, a small syntactic change (the lowercase versions of the primitives are allowed to have generic notation after them), and a bit of translation code in the parser, and we've already saved one Mars orbiter :)

The place where it would really shine would be when it gets first-class language support, though, because then you could allow for super-easy definition of units and even things like:

```csharp
public static double<T * U> RectangleArea(double<T> width, double<U> height)
        where T : INumericUnit
        where U : INumericUnit
    => width * height;
public static double<T> RectanglePerimeter(double<T> width, double<T> height)
        where T : INumericUnit
    => width * 2 + height * 2;

// "unit" implies readonly and, like "record", causes automatic method/field generation
public static unit struct SheepUnit : INumericUnit { }
// using * and / with unit types causes the creation of temporary types, like ValueTuple
public static unit struct AcreUnit = Meter * Meter * 4046.8;
// number of fenceposts isn't unitless, it's just it's own unit, this is equivalent to the SheepUnit declaration above
public static unit struct FencePostUnit = 1;

// in a "unit type" context, the "Unit" suffix is optional, like with Attributes
double<Sheep / Acre> sheepPerAcre = 10; // constant literals may be implicitly promoted to an absolute unit
double<Meter / FencePost> metersPerFencePost = 2;

double<Sheep / FencePost> SheepPerFencePost(double<Meter> pastureWidth, double<Meter> pastureHeight)
{
    // typecast to bare (int) preserves the unit if assigned to a unitful variable, it's the same semantic state as a constant literal
    int<Sheep> sheep = (int)Math.Floor(sheepPerAcre * RectangleArea(pastureWidth, pastureHeight)); // works
    int<FencePost> fenceposts = (int)Math.Ceil(RectangleArea(pastureWidth, pastureHeight) / metersPerFencePost); // oops, I meant RectanglePerimeter!
    // the above line fails to compile with ERROR: cannot convert int<FencePost * Meter> to int<FencePost>
    return (double)sheep / fenceposts;
}
```
